lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,917 @@
+  }
+
+  void addLabelReference(const std::list<parser::Label> &labels) {
+    for (const parser::Label &label : labels) {
+      addLabelReference(label);
+    }
+  }
+
+  std::vector<UnitAnalysis> programUnits_;
+  parser::Messages errorHandler_;
+  parser::CharBlock currentPosition_{nullptr};
+  ProxyForScope currentScope_{0};
+  std::vector<std::string> constructNames_;
+};
+
+bool InInclusiveScope(const std::vector<ProxyForScope> &scopes,
+    ProxyForScope tail, ProxyForScope head) {
+  for (; tail != head; tail = scopes[tail]) {
+    if (!HasScope(tail)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+ParseTreeAnalyzer LabelAnalysis(const parser::Program &program) {
+  ParseTreeAnalyzer analysis;
+  Walk(program, analysis);
+  return analysis;
+}
+
+bool InBody(const parser::CharBlock &position,
+    const std::pair<parser::CharBlock, parser::CharBlock> &pair) {
+  if (position.begin() >= pair.first.begin()) {
+    if (position.begin() < pair.second.end()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+LabeledStatementInfoTuplePOD GetLabel(
+    const TargetStmtMap &labels, const parser::Label &label) {
+  const auto iter{labels.find(label)};
+  if (iter == labels.cend()) {
+    return {0u, nullptr, LabeledStmtClassificationSet{}};
+  } else {
+    return iter->second;
+  }
+}
+
+// 11.1.7.3
+void CheckBranchesIntoDoBody(const SourceStmtList &branches,
+    const TargetStmtMap &labels, const std::vector<ProxyForScope> &scopes,
+    const IndexList &loopBodies, parser::Messages &errorHandler) {
+  for (const auto branch : branches) {
+    const auto &label{branch.parserLabel};
+    auto branchTarget{GetLabel(labels, label)};
+    if (HasScope(branchTarget.proxyForScope)) {
+      const auto &fromPosition{branch.parserCharBlock};
+      const auto &toPosition{branchTarget.parserCharBlock};
+      for (const auto body : loopBodies) {
+        if (!InBody(fromPosition, body) && InBody(toPosition, body)) {
+          errorHandler.Say(fromPosition,
+              parser::MessageFormattedText{
+                  "branch into '%s' from another scope"_en_US,
+                  body.first.ToString().c_str()});
+        }
+      }
+    }
+  }
+}
+
+void CheckDoNesting(
+    const IndexList &loopBodies, parser::Messages &errorHandler) {
+  for (auto i1{loopBodies.cbegin()}; i1 != loopBodies.cend(); ++i1) {
+    const auto &v1{*i1};
+    for (auto i2{i1 + 1}; i2 != loopBodies.cend(); ++i2) {
+      const auto &v2{*i2};
+      if (v2.first.begin() < v1.second.end() &&
+          v1.second.begin() < v2.second.begin()) {
+        errorHandler.Say(v2.second,
+            parser::MessageFormattedText{"'%s' doesn't properly nest"_err_en_US,
+                v1.first.ToString().c_str()});
+      }
+    }
+  }
+}
+
+parser::CharBlock SkipLabel(const parser::CharBlock &position) {
+  const long maxPosition{position.end() - position.begin()};
+  if (maxPosition && (position[0] >= '0') && (position[0] <= '9')) {
```

Comment:
Consider using `parser::CharBlock::empty()` and `parser::CharBlock::size()`.
Consider using `parser::IsDecimalDigit()`.